File: Zeze/Util/EventDispatcher.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Zeze.Util
{
    public class EventDispatcher
    {
        public string Name { get; }

        public EventDispatcher(string name)
        {
            this.Name = name;
        }

        public enum Mode
        {
            RunEmbed,
            RunProcedure,
            RunThread,
        }

        public class Events
        {
            private ConcurrentDictionary<long, Func<object, EventArgs, Task>> Handles = new();
            private Util.AtomicLong NextId = new();

            public class Canceler
            {
                public Events Events { get; }
                public long Id { get; }

                public Canceler(Events events, long id)
                {
                    Events = events;
                    Id = id;
                }

                public void Cancel()
                {
                    Events.Handles.Remove(Id, out _);
                }
            }

            public Canceler Add(Func<object, EventArgs, Task<long>> handle)
            {
                var next = NextId.IncrementAndGet();
                if (!Handles.TryAdd(next, handle))
                {
                    throw new Exception("Impossible!");
                }
                return new Canceler(this, next);
            }

            public ICollection<Func<object, EventArgs, Task>> values()
            {
                return this.Handles.Values;
            }
        }

        public Events RunEmbedEvents { get; } = new();
        public Events RunProcedureEvents { get; } = new();
        public Events RunThreadEvents { get; } = new();

        /**
         * 注册事件处理函数，
         * @return 如果需要取消注册，请保存返回值，并调用其cancel。
         */
        public Events.Canceler Add(Mode mode, Func<object, EventArgs, Task<long>> handle)
        {
            switch (mode)
            {
                case Mode.RunEmbed:
                    return RunEmbedEvents.Add(handle);

                case Mode.RunProcedure:
                    return RunProcedureEvents.Add(handle);

                case Mode.RunThread:
                    return RunThreadEvents.Add(handle);
            }
            throw new Exception($"Unknown mode={mode}");
        }

        // 事件派发。需要触发者在明确的地方显式的调用。

        // 启动新的线程执行。
        public void TriggerThread(object sender, EventArgs arg)
        {
            foreach (var handle in RunThreadEvents.values())
            {
                _ = Mission.CallAsync(async () =>
                {
                    await handle(sender, arg);
                    return 0;
                }, $"EventDispatch.{Name}.TriggerThread");
            }
        }

        private static readonly ILogger logger = LogManager.GetLogger(typeof(EventDispatcher));

        // 在当前线程中，创建新的存储过程并执行，忽略所有错误。
        public async Task TriggerProcedureIgnoreError(Application app, object sender, EventArgs arg)
        {
            foreach (var handle in RunProcedureEvents.values())
            {
                try
                {
                    await app.NewProcedure(async () =>
                    {
                        await handle(sender, arg);
                        return 0;
                    }, $"EventDispatch.{Name}.TriggerProcedureIgnoreError").CallAsync();
                    // return error has log inner.
                }
                catch (Exception ex)
                {
                    logger.Error(ex);
                }
            }
        }

        // 嵌入当前线程执行，所有错误都报告出去，如果需要对错误进行特别处理，需要自己遍历Handles手动触发。
        public async Task TriggerEmbed(object sender, EventArgs arg)
        {
            foreach (var handle in RunEmbedEvents.values())
            {
                await handle(sender, arg);
            }
        }

        // 在当前线程中，创建新的存储过程并执行，所有错误都报告出去，如果需要对错误进行特别处理，需要自己遍历Handles手动触发。
        public async Task TriggerProcedure(Application app, object sender, EventArgs arg)
        {
            foreach (var handle in RunProcedureEvents.values())
            {
                var rc = await app.NewProcedure(async () =>
                {
                    await handle(sender, arg);
                    return 0L;
                }, $"EventDispatch.{Name}.TriggerProcedure").CallAsync();
                if (0L != rc)
                    throw new Exception($"Nest Call Fail. return={rc}");
            }
        }
    }
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Util/EventDispatcher.java
package Zeze.Util;

import java.util.concurrent.ConcurrentLinkedQueue;
import Zeze.Application;
import Zeze.Transaction.DispatchMode;
import org.jetbrains.annotations.NotNull;

public class EventDispatcher {
	// private static final @NotNull Logger logger = LogManager.getLogger(EventDispatcher.class);

	@FunctionalInterface
	public interface EventHandle {
		long invoke(@NotNull Object sender, @NotNull EventArgument arg) throws Exception;
	}

	public interface EventArgument {
	}

	public interface Canceler {
		void cancel();
	}

	public enum Mode {
		RunEmbed,
		RunProcedure,
		RunThread,
	}

	static class HandleClass {
		final @NotNull Object classNameOrHandle; // EventHandle handle || String className

		HandleClass(@NotNull EventHandle handle) {
			this.classNameOrHandle = handle;
		}

		HandleClass(@NotNull String className) {
			this.classNameOrHandle = className;
		}
	}

	private final @NotNull Application zeze;
	private final @NotNull String name;
	private final ConcurrentLinkedQueue<HandleClass> runEmbedEvents = new ConcurrentLinkedQueue<>();
	private final ConcurrentLinkedQueue<HandleClass> runProcedureEvents = new ConcurrentLinkedQueue<>();
	private final ConcurrentLinkedQueue<HandleClass> runThreadEvents = new ConcurrentLinkedQueue<>();

	public EventDispatcher(@NotNull Application zeze, @NotNull String name) {
		this.zeze = zeze;
		this.name = name;
	}

	public @NotNull String getName() {
		return name;
	}

	private @NotNull ConcurrentLinkedQueue<HandleClass> getQueue(@NotNull Mode mode) {
		switch (mode) {
		case RunEmbed:
			return runEmbedEvents;
		case RunProcedure:
			return runProcedureEvents;
		case RunThread:
			return runThreadEvents;
		default:
			throw new IllegalArgumentException("Unknown mode=" + mode);
		}
	}

	/**
	 * 注册事件处理函数，如果在事件触发过程中执行注册,则等下次再触发
	 *
	 * @return 如果需要取消注册，请保存返回值，并调用其cancel。
	 */
	public @NotNull Canceler addHot(@NotNull Mode mode, @NotNull Class<? extends EventHandle> handleClass) {
		var events = getQueue(mode);
		var handle = new HandleClass(handleClass.getName());
		events.offer(handle);
		return () -> events.remove(handle);
	}

	/**
	 * 注册事件处理函数，如果在事件触发过程中执行注册,则等下次再触发
	 *
	 * @return 如果需要取消注册，请保存返回值，并调用其cancel。
	 */
	public @NotNull Canceler add(@NotNull Mode mode, @NotNull EventHandle handle_) {
		zeze.verifyCallerCold(StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE).getCallerClass());
		var events = getQueue(mode);
		var handle = new HandleClass(handle_);
		events.offer(handle);
		return () -> events.remove(handle);
	}

	// 事件派发。需要触发者在明确的地方显式的调用。

	// 启动新的线程执行。
	public void triggerThread(@NotNull Object sender, @NotNull EventArgument arg) {
		for (var handle : runThreadEvents) {
			var classNameOrHandle = handle.classNameOrHandle;
			if (classNameOrHandle instanceof EventHandle) {
				Task.run(() -> ((EventHandle)classNameOrHandle).invoke(sender, arg),
						"EventDispatch." + name + ".runAsync", DispatchMode.Normal);
			} else {
				var className = (String)classNameOrHandle;
				Task.run(() -> zeze.getHotHandle().findHandle(zeze, className).invoke(sender, arg),
						"EventDispatch." + name + ".runAsync " + className, DispatchMode.Normal);
			}
		}
	}

	// 嵌入当前线程执行，所有错误都报告出去，如果需要对错误进行特别处理，需要自己遍历Handles手动触发。
	public long triggerEmbed(@NotNull Object sender, EventArgument arg) throws Exception {
		for (var handle : runEmbedEvents) {
			var classNameOrHandle = handle.classNameOrHandle;
			var ret = classNameOrHandle instanceof EventHandle
					? ((EventHandle)classNameOrHandle).invoke(sender, arg)
					: zeze.getHotHandle().findHandle(zeze, (String)classNameOrHandle).invoke(sender, arg);
			if (ret != 0)
				return ret;
		}
		return 0;
	}

	// 在当前线程中，创建新的存储过程并嵌套执行，所有错误都报告出去，如果需要对错误进行特别处理，需要自己遍历Handles手动触发。
	public void triggerProcedure(@NotNull Application app, @NotNull Object sender, @NotNull EventArgument arg) {
		for (var handle : runProcedureEvents) {
			var classNameOrHandle = handle.classNameOrHandle;
			if (classNameOrHandle instanceof EventHandle) {
				Task.call(app.newProcedure(() -> {
					((EventHandle)classNameOrHandle).invoke(sender, arg); // 忽略嵌套的存储的执行。
					return 0L;
				}, "EventDispatcher.triggerProcedure"));
			} else {
				var className = (String)classNameOrHandle;
				Task.call(app.newProcedure(() -> {
					zeze.getHotHandle().findHandle(zeze, className).invoke(sender, arg); // 忽略嵌套的存储的执行。
					return 0L;
				}, "EventDispatcher.triggerProcedure." + className));
			}
		}
	}
}
